src/index.tsx

```diff
@@ -51,32 +52,79 @@ class CommLock {
   promise: any;
   requestedLockCount: number;
   clearedLockCount: number;
-
-  constructor() {
+  _activeTimeout: number;
+  _lockTimeout: number;
+  _timeoutSignal: Signal<CommLock, number>;
+  _recentKernelState: string;
+  _sessionContext: ISessionContext;
+
+  constructor(_lockTimeout: number, sessionContext: ISessionContext) {
+    this._lockTimeout = _lockTimeout;
+    this._activeTimeout = null;
     this.requestedLockCount = 0;
     this.clearedLockCount = 0;
     this._disable = {};
     this.promise = { 0: Promise.resolve() };
+    this._timeoutSignal = new Signal<CommLock, number>(this);
+    this._sessionContext = sessionContext;
+    this._sessionContext.statusChanged.connect(this.kernelStateRecorder, this);
+    this._timeoutSignal.connect(this.timeoutExpireHandler, this);
+  }
+
+  kernelStateRecorder(sender: ISessionContext, args: Kernel.Status) {
+    this._recentKernelState = args;
+  }
+
+  _clearTimeout() {
+    window.clearTimeout(this._activeTimeout);
+    this._activeTimeout = null;
+  }
+
+  /*
+		If the kernel was busy the last time, we assume it was busy executing 
+		code and we restart the timeout.
+  */
+  timeoutExpireHandler(sender: CommLock, id: number) {
+    this._clearTimeout();
+    if (this._recentKernelState === 'busy') {
+      console.debug('Extending Timeout For: ', id);
+      this.createTimeout(id);
+    } else {
+      this.release(id);
+    }
+  }
+
+  async acquire(): Promise<number> {
+    const lockId = this.requestedLockCount++;
```

Comment:
If I understand this section right, I think it can be better written as:

```
const lastLockPromise = this.promise[this.requestedLockCount];
this.requestedLockCount++;
const lockId = this.requestedLockCount;
    this.promise[lockId] = new Promise(resolve => {
      this._disable[lockId] = resolve;
    });
await lastLockPromise;
return new Promise((res, rej) => res(lockId));
```